airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py

```diff
@@ -161,50 +162,63 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
-                                    yield transform(out_message)
-                                else:
-                                    # todo: check that messages match the discovered schema
-                                    stream_name = transformed_json["stream"]
-                                    out_record = AirbyteRecordMessage(
-                                        stream=stream_name,
-                                        data=transformed_json["record"],
-                                        emitted_at=int(datetime.now().timestamp()) * 1000,
-                                    )
-                                    out_message = AirbyteMessage(type=Type.RECORD, record=out_record)
-                                    yield transform(out_message)
-                        else:
-                            logger.log_by_prefix(line, "INFO")
+            for line, text_wrapper in SingerHelper._read_lines(p):
+                if text_wrapper is p.stdout:
+                    out_json = to_json(line)
+                    if out_json is not None and is_message(out_json):
+                        message_data = SingerHelper._airbyte_message_from_json(out_json)
+                        if message_data is not None:
+                            yield message_data
                     else:
-                        logger.log_by_prefix(line, "ERROR")
+                        logger.log_by_prefix(line, "INFO")
+                else:
+                    logger.log_by_prefix(line, "ERROR")
+
+    @staticmethod
+    def _read_lines(process: subprocess.Popen) -> Iterator[Tuple[str, TextIOWrapper]]:
+        sel = selectors.DefaultSelector()
+        sel.register(process.stdout, selectors.EVENT_READ)
+        sel.register(process.stderr, selectors.EVENT_READ)
+        eof = False
+        while not eof:
+            selects_list = sel.select()
+            empty_line_counter = 0
+            for key, _ in selects_list:
+                line = key.fileobj.readline()
+                if not line:
+                    empty_line_counter += 1
+                    if empty_line_counter >= len(selects_list):
+                        eof = True
+
+                    try:
+                        process.wait(timeout=60)
+                    except subprocess.TimeoutExpired:
+                        raise Exception(f"Underlying command {process.args} is hanging")
+
+                    if process.returncode != 0:
+                        raise Exception(f"Underlying command {process.args} failed with exit code {process.returncode}")
+                else:
+                    yield line, key.fileobj
+
+    @staticmethod
+    def _airbyte_message_from_json(transformed_json: Mapping[str, Any]) -> Optional[AirbyteMessage]:
+        if transformed_json is None or transformed_json.get("type") == "SCHEMA" or transformed_json.get("type") == "ACTIVATE_VERSION":
```

Comment:
`if transformed_json is None or transformed_json.get("type") in {"SCHEMA", "ACTIVATE_VERSION"}: